project5/3700kvstore

```diff
@@ -78,215 +80,255 @@ class Server:
         self.match_idx = {} # index of highest log entry known on each server
         self.prev_heartbeat = current_time() # timestamp of previous heartbeat
 
-        # Start heartbeat thread
-        threading.Thread(target=self.send_heartbeat).start()
+        self.state = None
+        self.transition_to_state(State.CANDIDATE)
+
+    @property
+    def last_log_idx(self):
+        return len(self.log) - 1
+        
+    @property
+    def quorum(self):
+        return len(self.replicas)//2 + 1
+    
+    def transition_to_state(self, state, msg=None, term=None):
+        if self.state != state:
+            logging.info('Term {} {} -> {}'.format(str(self.current_term),
+                str(self.state), str(state)))
+
+        if state == State.LEADER:
+            logging.info('Voted for by {}'.format(str(list(self.votes))))
+            # Re-initialize any necessary fields
+            self.state = State.LEADER
+            self.leader = self.id
+            self.prev_heartbeat = current_time()
+            for replica in self.replicas:
+                # For each server, index of next log entry to send to that server
+                # (initialized to leader last log index + 1)
+                self.next_idx[replica] = self.commit_idx + 1
+                # For each server, index of highest log entry known to be 
+                # replicated on server (initialized to 0, increases monotonically)
+                self.match_idx[replica] = 0
+
+            self.broadcast_heartbeat()
+            
+        elif state == State.FOLLOWER:
+            self.state = State.FOLLOWER
+            self.leader = msg['leader']
+            self.current_term = msg['term']
+            self.votes = set()
+            self.voted_for = None
+
+        elif state == State.CANDIDATE:
+            self.state = State.CANDIDATE
+            self.leader = 'FFFF' # ID of the assumed leader
+            self.reset_timeouts()
+            if term is None:
+                self.current_term += 1
+            else:
+                self.current_term = term
+            # Reset any votes from a previous election
+            self.votes = set((self.id,))
```

Comment:
What does the double parentheses do? And did you mean to leave that comma there?

Reply:
The parens + comma make it a tuple with one element. Just parens only affects order of operations, and comma outside of the parens is a malformed argument list

Reply:
Gotcha. Thanks!